File: apps/scanner/app/src/main/java/com/airferry/app/scan/Af2LedgerStore.kt
package com.airferry.app.scan

import java.io.File
import java.io.FileOutputStream
import org.json.JSONObject

/**
 * Crash-safe §12 resume ledger — the journal twin of [ChunkSpillStore]'s
 * `.partial` file.
 *
 * Format: one JSONL journal per transfer, `af2-<tid>.ledger.jsonl`.
 * - Line 1 (header): `{"v":1,"tid":…,"root":…,"crs":…}` — written once,
 *   atomically (temp file + fsync + rename) before the first chunk commit.
 * - Each later line: `{"c":<index>}` (chunk committed after spill+fsync) or
 *   `{"i":<index>}` (chunk invalidated after a re-verification failure).
 *
 * Per-line append + fsync keeps every interleaving crash-safe: a torn tail
 * line fails `JSONObject` parsing and is skipped, so the ledger never
 * reports MORE than what hit the disk. A commit line is only appended after
 * the chunk bytes were pwrite + fsync'd into the spill — the §12 ordering
 * rule "账本完成 ⇒ 数据已落盘" holds by construction.
 *
 * Only touched from the ingest thread (the decode pool serializes ingest)
 * and the recovery path that runs under the same lock.
 */
class Af2LedgerStore private constructor(private val path: File) {

    var transferIdHex: String = ""
        private set
    var chunkRawSize: Int = 0
        private set
    /** ROOT frame bytes (hex at rest) for the §12 resume() call. */
    var rootFrameBytes: ByteArray = ByteArray(0)
        private set
    private val completed = sortedSetOf<Int>()

    val completedIndices: IntArray get() = completed.toIntArray()

    /** Load an existing ledger for `tid`, or null when none exists. */
    fun reload(): Boolean {
        completed.clear()
        if (!path.isFile) return false
        val lines = try {
            path.readLines()
        } catch (_: Exception) {
            return false
        }
        var header: JSONObject? = null
        for (line in lines) {
            if (line.isBlank()) continue
            val o = try {
                JSONObject(line)
            } catch (_: Exception) {
                continue // torn tail line from a mid-write crash
            }
            if (header == null && o.has("v")) {
                header = o
                continue
            }
            if (o.has("c")) completed.add(o.getInt("c"))
            if (o.has("i")) completed.remove(o.getInt("i"))
        }
        val h = header ?: return false
        transferIdHex = h.optString("tid", "")
        chunkRawSize = h.optInt("crs", 0)
        rootFrameBytes = hexToBytes(h.optString("root", ""))
        return transferIdHex.isNotEmpty() && rootFrameBytes.isNotEmpty()
    }

    /** Append one commit event (after the chunk was spilled + fsync'd). */
    fun commit(index: Int) {
        appendLine(JSONObject().put("c", index))
        completed.add(index)
    }

    /** Append one invalidate event (after a spill re-verification failure). */
    fun invalidate(index: Int) {
        appendLine(JSONObject().put("i", index))
        completed.remove(index)
    }

    private fun appendLine(o: JSONObject) {
        try {
            FileOutputStream(path, true).use { fos ->
                fos.write((o.toString() + "\n").toByteArray())
                fos.fd.sync()
            }
        } catch (e: Exception) {
            android.util.Log.w("Af2LedgerStore", "append failed", e)
        }
    }

    /** Delete the journal (transfer finished / relocked away / abandoned). */
    fun discard() {
        path.delete()
    }

    data class PendingTransfer(
        val transferIdHex: String,
        val chunkRawSize: Int,
        val completedCount: Int,
        val diskBytes: Long,
        val lastModified: Long,
    )

    companion object {
        /** List all uncompleted/partial transfer ledgers in `dir`. */
        fun listPendingTransfers(dir: File): List<PendingTransfer> {
            val candidates = dir.listFiles { f -> f.name.startsWith("af2-") && f.name.endsWith(".ledger.jsonl") }
                ?: return emptyList()
            val list = mutableListOf<PendingTransfer>()
            for (f in candidates) {
                val store = Af2LedgerStore(f)
                if (store.reload()) {
                    val tid = store.transferIdHex
                    val spill = File(dir, "af2-$tid.partial")
                    val spillBytes = if (spill.isFile) spill.length() else 0L
                    list.add(
                        PendingTransfer(
                            transferIdHex = tid,
                            chunkRawSize = store.chunkRawSize,
                            completedCount = store.completedIndices.size,
                            diskBytes = f.length() + spillBytes,
                            lastModified = f.lastModified()
                        )
                    )
                }
            }
            return list.sortedByDescending { it.lastModified }
        }

        /** Discard all pending journals and spill files in `dir`. */
        fun discardAllPending(dir: File) {
            dir.listFiles { f ->
                f.name.startsWith("af2-") && (
                    f.name.endsWith(".ledger.jsonl") ||
                    f.name.endsWith(".partial") ||
                    f.name.endsWith(".tmp")
                )
            }?.forEach { it.delete() }
        }

        /** Resume source: newest valid ledger in `dir` (by mtime), or null. */
        fun loadMostRecent(dir: File): Af2LedgerStore? {
            val candidates = dir.listFiles { f -> f.name.endsWith(".ledger.jsonl") }
                ?: return null
            for (candidate in candidates.sortedByDescending { it.lastModified() }) {
                val store = Af2LedgerStore(candidate)
                if (store.reload()) return store
            }
            return null
        }

        /** Remove unrecoverable spill files that have no valid resume journal. */
        fun sweepOrphanPartials(dir: File) {
            val validTids = mutableSetOf<String>()
            dir.listFiles { f -> f.name.endsWith(".ledger.jsonl") }?.forEach { file ->
                val store = Af2LedgerStore(file)
                if (store.reload()) {
                    validTids.add(store.transferIdHex)
                } else {
                    file.delete()
                }
            }
            dir.listFiles { f -> f.name.startsWith("af2-") && f.name.endsWith(".partial") }
                ?.forEach { partial ->
                    val tid = partial.name.removePrefix("af2-").removeSuffix(".partial")
                    if (tid !in validTids) partial.delete()
                }
        }

        /** Create + write the header for a fresh transfer's journal. */
        fun create(
            dir: File,
            transferIdHex: String,
            chunkRawSize: Int,
            rootFrameBytes: ByteArray
        ): Af2LedgerStore {
            val path = File(dir, "af2-${transferIdHex.ifEmpty { "session" }}.ledger.jsonl")
            path.delete() // a relock restarts the journal from scratch
            val header = JSONObject()
                .put("v", 1)
                .put("tid", transferIdHex)
                .put("crs", chunkRawSize)
                .put("root", bytesToHex(rootFrameBytes))
            // Atomic header: temp + fsync + rename so a crash mid-create
            // never leaves a headerless journal that a later commit would
            // append to.
            val tmp = File(dir, path.name + ".tmp")
            try {
                FileOutputStream(tmp).use { fos ->
                    fos.write((header.toString() + "\n").toByteArray())
                    fos.fd.sync()
                }
                if (!tmp.renameTo(path)) {
                    // Cross-filesystem rename fallback (cache dirs are same-FS
                    // in practice; copy keeps the fsync-before-rename order).
                    tmp.copyTo(path, overwrite = true)
                    path.setLastModified(System.currentTimeMillis())
                    tmp.delete()
                }
            } catch (e: Exception) {
                android.util.Log.w("Af2LedgerStore", "header write failed", e)
                tmp.delete()
            }
            return Af2LedgerStore(path).apply {
                this.transferIdHex = transferIdHex
                this.chunkRawSize = chunkRawSize
                this.rootFrameBytes = rootFrameBytes
            }
        }

        /** Drop the journal (and spill) of `tid` — completion cleanup. */
        fun discardFor(dir: File, transferIdHex: String) {
            File(dir, "af2-$transferIdHex.ledger.jsonl").delete()
        }

        private fun bytesToHex(b: ByteArray): String =
            b.joinToString("") { "%02x".format(it) }

        private fun hexToBytes(s: String): ByteArray {
            if (s.length % 2 != 0) return ByteArray(0)
            return ByteArray(s.length / 2) { i ->
                ((Character.digit(s[i * 2], 16) shl 4) +
                    Character.digit(s[i * 2 + 1], 16)).toByte()
            }
        }
    }
}

File: apps/windows/AirFerry.Windows/Scan/Af2LedgerStore.cs
using System.IO;
using System.Text;
using System.Text.Json;

namespace AirFerry.Windows.Scan;

/// <summary>
/// Crash-safe §12 resume ledger — the journal twin of <see cref="ChunkSpillStore"/>'s
/// <c>.partial</c> file. JSONL, one file per transfer (<c>af2-&lt;tid&gt;.ledger.jsonl</c>):
/// <para>
/// Line 1 (header): <c>{"v":1,"tid":…,"root":…,"crs":…}</c> — written atomically
/// (temp + flush + rename) before the first chunk commit. Each later line is
/// <c>{"c":i}</c> (chunk committed after its bytes were pwrite+fsync'd into the
/// spill) or <c>{"i":i}</c> (chunk invalidated after a re-verification failure).
/// A torn tail line fails JSON parsing and is skipped, so the journal never
/// reports more than what reached the disk.
/// </para>
/// <para>
/// Only touched from the pool's serialized ingest callback (under
/// <c>IngestLock</c>) and the recovery core that runs under the same lock.
/// </para>
/// </summary>
public sealed class Af2LedgerStore
{
    private readonly string _path;
    public string TransferIdHex { get; private set; } = "";
    public int ChunkRawSize { get; private set; }
    public byte[] RootFrameBytes { get; private set; } = Array.Empty<byte>();
    public SortedSet<int> Completed { get; } = new();

    private Af2LedgerStore(string path)
    {
        _path = path;
    }

    /// <summary>True once the header line is durably on disk (set by Create).</summary>
    private bool _headerDurable;

    public int[] CompletedIndices => Completed.ToArray();

    /// <summary>Parse (or re-parse) the journal. True when a valid header exists.</summary>
    public bool Reload()
    {
        Completed.Clear();
        if (!File.Exists(_path))
        {
            return false;
        }
        string[] lines;
        try
        {
            lines = File.ReadAllLines(_path);
        }
        catch (IOException)
        {
            return false;
        }
        bool headerSeen = false;
        foreach (string line in lines)
        {
            if (string.IsNullOrWhiteSpace(line))
            {
                continue;
            }
            JsonElement o;
            try
            {
                using JsonDocument doc = JsonDocument.Parse(line);
                o = doc.RootElement.Clone();
            }
            catch (JsonException)
            {
                continue; // torn tail line from a mid-write crash
            }
            if (!headerSeen)
            {
                if (!o.TryGetProperty("v", out _))
                {
                    continue;
                }
                TransferIdHex = o.TryGetProperty("tid", out JsonElement tid) ? tid.GetString() ?? "" : "";
                ChunkRawSize = o.TryGetProperty("crs", out JsonElement crs) && crs.TryGetInt32(out int crsVal)
                    ? crsVal : 0;
                RootFrameBytes = o.TryGetProperty("root", out JsonElement root)
                    ? HexToBytes(root.GetString() ?? "") : Array.Empty<byte>();
                headerSeen = true;
                continue;
            }
            if (o.TryGetProperty("c", out JsonElement c) && c.TryGetInt32(out int ci))
            {
                Completed.Add(ci);
            }
            if (o.TryGetProperty("i", out JsonElement inv) && inv.TryGetInt32(out int ii))
            {
                Completed.Remove(ii);
            }
        }
        // A journal that reloads with a valid header is durable by definition —
        // resumed transfers keep appending to it.
        _headerDurable = headerSeen &&
            !string.IsNullOrEmpty(TransferIdHex) &&
            RootFrameBytes.Length > 0;
        return _headerDurable;
    }

    /// <summary>Append one commit event (after the chunk was spilled + flushed).</summary>
    public void Commit(int index)
    {
        if (!_headerDurable) return; // headerless journal would never reload
        AppendLine($"{{\"c\":{index}}}");
        Completed.Add(index);
    }

    /// <summary>Append one invalidate event (after a re-verification failure).</summary>
    public void Invalidate(int index)
    {
        if (!_headerDurable) return;
        AppendLine($"{{\"i\":{index}}}");
        Completed.Remove(index);
    }

    private void AppendLine(string json)
    {
        try
        {
            using var fs = new FileStream(
                _path, FileMode.Append, FileAccess.Write, FileShare.None);
            byte[] bytes = Encoding.UTF8.GetBytes(json + "\n");
            fs.Write(bytes, 0, bytes.Length);
            fs.Flush(flushToDisk: true);
        }
        catch (Exception ex)
        {
            System.Diagnostics.Debug.WriteLine($"[Af2LedgerStore] append failed: {ex.Message}");
        }
    }

    /// <summary>Delete the journal (transfer finished / relocked away / abandoned).</summary>
    public void Discard()
    {
        try { File.Delete(_path); } catch (IOException) { }
    }

    public record PendingTransfer(
        string TransferIdHex,
        int ChunkRawSize,
        int CompletedCount,
        long DiskBytes,
        DateTime LastModified);

    /// <summary>List all pending/uncompleted transfer ledgers in <paramref name="dir"/>.</summary>
    public static IReadOnlyList<PendingTransfer> ListPendingTransfers(string dir)
    {
        if (!Directory.Exists(dir)) return [];
        try
        {
            var list = new List<PendingTransfer>();
            foreach (var file in Directory.EnumerateFiles(dir, "af2-*.ledger.jsonl"))
            {
                var store = new Af2LedgerStore(file);
                if (store.Reload())
                {
                    string tid = store.TransferIdHex;
                    string spill = Path.Combine(dir, $"af2-{tid}.partial");
                    long spillBytes = File.Exists(spill) ? new FileInfo(spill).Length : 0L;
                    var fi = new FileInfo(file);
                    list.Add(new PendingTransfer(
                        tid,
                        store.ChunkRawSize,
                        store.CompletedIndices.Length,
                        fi.Length + spillBytes,
                        fi.LastWriteTime));
                }
            }
            return list.OrderByDescending(p => p.LastModified).ToList();
        }
        catch (Exception)
        {
            return [];
        }
    }

    /// <summary>Discard all pending journals, spills, and temp files in <paramref name="dir"/>.</summary>
    public static void DiscardAllPending(string dir)
    {
        if (!Directory.Exists(dir)) return;
        try
        {
            foreach (var f in Directory.EnumerateFiles(dir, "af2-*"))
            {
                if (f.EndsWith(".ledger.jsonl") || f.EndsWith(".partial") || f.EndsWith(".tmp"))
                {
                    try { File.Delete(f); } catch { }
                }
            }
        }
        catch { }
    }

    /// <summary>Resume source: the newest valid journal in <paramref name="dir"/>.</summary>
    public static Af2LedgerStore? LoadMostRecent(string dir)
    {
        try
        {
            if (!Directory.Exists(dir))
            {
                return null;
            }
            foreach (FileInfo candidate in new DirectoryInfo(dir)
                         .EnumerateFiles("*.ledger.jsonl")
                         .OrderByDescending(f => f.LastWriteTimeUtc))
            {
                var store = new Af2LedgerStore(candidate.FullName);
                if (store.Reload()) return store;
            }
            return null;
        }
        catch (Exception)
        {
            return null;
        }
    }

    /// <summary>Remove spill files that have no parseable resume journal.</summary>
    public static void SweepOrphanPartials(string dir)
    {
        if (!Directory.Exists(dir)) return;
        try
        {
            var validTids = new HashSet<string>(StringComparer.Ordinal);
            foreach (string journal in Directory.EnumerateFiles(dir, "*.ledger.jsonl"))
            {
                var store = new Af2LedgerStore(journal);
                if (store.Reload())
                {
                    validTids.Add(store.TransferIdHex);
                }
                else
                {
                    try { File.Delete(journal); } catch { }
                }
            }
            foreach (string partial in Directory.EnumerateFiles(dir, "af2-*.partial"))
            {
                string name = Path.GetFileName(partial);
                string tid = name.Substring(4, name.Length - 4 - ".partial".Length);
                if (!validTids.Contains(tid))
                {
                    try { File.Delete(partial); } catch { }
                }
            }
        }
        catch { }
    }

    /// <summary>Create + atomically write the header for a fresh transfer's journal.</summary>
    public static Af2LedgerStore Create(
        string dir, string transferIdHex, int chunkRawSize, byte[] rootFrameBytes)
    {
        string id = string.IsNullOrEmpty(transferIdHex) ? "session" : transferIdHex;
        string path = Path.Combine(dir, $"af2-{id}.ledger.jsonl");
        try { File.Delete(path); } catch (IOException) { }
        string header = JsonSerializer.Serialize(new
        {
            v = 1,
            tid = transferIdHex,
            crs = chunkRawSize,
            root = BytesToHex(rootFrameBytes),
        });
        bool headerDurable = false;
        try
        {
            Directory.CreateDirectory(dir);
            string tmp = path + ".tmp";
            File.WriteAllText(tmp, header + "\n");
            File.Move(tmp, path, overwrite: true);
            headerDurable = true;
        }
        catch (Exception ex)
        {
            // A journal whose header never became durable is permanently
            // un-reloadable (Reload requires a header line) and the resume
            // sweep would delete its spill as garbage. Keep the store for the
            // in-memory completed-set, but refuse to append so we never write
            // a headerless {"c":i} journal that silently breaks §12 resume.
            System.Diagnostics.Debug.WriteLine($"[Af2LedgerStore] header write failed: {ex.Message}");
        }
        return new Af2LedgerStore(path)
        {
            TransferIdHex = transferIdHex,
            ChunkRawSize = chunkRawSize,
            RootFrameBytes = rootFrameBytes,
            _headerDurable = headerDurable,
        };
    }

    private static string BytesToHex(byte[] b)
    {
        var sb = new StringBuilder(b.Length * 2);
        foreach (byte x in b)
        {
            sb.Append(x.ToString("x2"));
        }
        return sb.ToString();
    }

    private static byte[] HexToBytes(string s)
    {
        if (s.Length == 0 || s.Length % 2 != 0)
        {
            return Array.Empty<byte>();
        }
        var outBytes = new byte[s.Length / 2];
        for (int i = 0; i < outBytes.Length; i++)
        {
            outBytes[i] = Convert.ToByte(s.Substring(i * 2, 2), 16);
        }
        return outBytes;
    }
}
